### PR TITLE
sql: Implement datediff

### LINKED ISSUE #19790: sql: Implement the datediff function
### Feature request

We should implement the [`datediff(...)`](https://docs.aws.amazon.com/redshift/latest/dg/r_DATEDIFF_function.html) function as it exists in RedShift. This function doesn't exist in Postgres, but it's been asked for by a number of folks and we've determined diverging from Postgres in this case is okay, [slack](https://materializeinc.slack.com/archives/CU7ELJ6E9/p1686161391338089).

### PR DESCRIPTION
**Pull request by ParkMyCar**
### Motivation

* Fixes #19790

This PR adds the [`datediff(...)`](https://docs.aws.amazon.com/redshift/latest/dg/r_DATEDIFF_function.html) function, which is not in Postgres, but it is in a number of other analytical databases, including RedShift.

For the test cases, I spun up a version of MSSQL in Docker and asserted that we return the same values they do.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - This PR adds the `datediff(...)` function which returns the number of datepart boundaries that are crossed between two dates.

### CODE REVIEW

**Review by jkosh44 — APPROVED** [review]

File: doc/user/content/sql/functions/datediff.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+```
+SELECT datediff('millennia', '2000-12-31', '2001-01-01') as d;
+  d
+-----
+  1
+```
+
+Even though the difference between `2000-12-31` and `2001-01-01` is only a single day, we cross
+the millenium boundary from one date to the other, hence the above statement returns 1.
```
Comment (morsapaes) [inline doc/user/content/sql/functions/datediff.md:32]:
```suggestion
Even though the difference between `2000-12-31` and `2001-01-01` is a single day, a millennium boundary is crossed from one date to the other, so the result is `1`.
```

File: doc/user/content/sql/functions/datediff.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+This function also considers leap years.
+
```
Comment (morsapaes) [inline doc/user/content/sql/functions/datediff.md:35]:
```suggestion

To see how this function handles leap years:
```

File: doc/user/content/sql/functions/datediff.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+     2
+
+SELECT datediff('day', '2005-02-28', '2005-03-01') as non_leap;
+  non_leap
+------------
+     1
+```
+
+In the first statement when using the year `2004` (a leap year), the number of day boundaries we
+cross is 2. Whereas in the second statement when we use the year `2005`, we only cross 1 day boundary.
```
Comment (morsapaes) [inline doc/user/content/sql/functions/datediff.md:49]:
```suggestion
In the statement that uses a leap year (`2004`), the number of day boundaries crossed is `2`. When using a non-leap year (`2005`), only `1` day boundary is crossed.
```

**Review by morsapaes — APPROVED** [review]

File: src/expr/src/scalar/func.rs [hunk 4]
```diff
@@ -7445,7 +7527,11 @@ impl VariadicFunc {
             | VariadicFunc::Greatest
             | VariadicFunc::Least
             | VariadicFunc::And
-            | VariadicFunc::Or => true,
+            | VariadicFunc::Or
+            | VariadicFunc::DateDiffTimestamp
+            | VariadicFunc::DateDiffTimestampTz
+            | VariadicFunc::DateDiffDate
+            | VariadicFunc::DateDiffTime => true,
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/expr/src/scalar/func.rs:7534]: Ah ha! Great catch, you're totally right, let me fix this

**Pull request merged**